supernode/daemon/mgr/progress/piece_state.go

```diff
@@ -0,0 +1,130 @@
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// pieceState maintains the information about
+// which peers the piece currently exists on.
+type pieceState struct {
+	superCID string
+
+	pieceContainer   map[string]bool
+	distributedCount int32
+	sync.Mutex
+}
+
+// newPieceState returns a new pieceState.
+func newPieceState() *pieceState {
+	return &pieceState{
+		pieceContainer: make(map[string]bool, 0),
+	}
+}
+
+// add a clientID for the corresponding piece which means that
+// there is a new peer node that owns this piece.
+func (ps *pieceState) add(clientID string) error {
+	if cutil.IsEmptyStr(clientID) {
+		return errors.Wrap(errorType.ErrEmptyValue, "clientID")
+	}
+
+	if isSupernode(clientID) {
+		if ps.superCID == "" {
+			// NOTE: should we use the new value or keep it?
+			ps.superCID = clientID
+		}
+		return nil
+	}
+
+	ps.Lock()
+	defer ps.Unlock()
+	if _, ok := ps.pieceContainer[clientID]; ok {
+		logrus.Warnf("clientID: %s is exist", clientID)
+		return nil
+	}
+
+	ps.pieceContainer[clientID] = true
+	ps.distributedCount = ps.distributedCount + 1
+	return nil
+}
+
+// getDstCID returns a available clientID.
+func (ps *pieceState) getDstCID(taskID, clientID string, pieceLen int, blackInfo map[string]bool,
+	clientErrCount int, serviceErrorInfo, serviceDownInfo, producerLoad cutil.SyncMap) (dstCID string) {
+
+	defer func() {
+		// use superCID as the default dstCID.
+		if cutil.IsEmptyStr(dstCID) && !cutil.IsEmptyStr(ps.superCID) {
+			// TODO: rate limit
+			dstCID = ps.superCID
+		}
+	}()
+
+	// don't dispatch to other peer nodes in the network
+	// if reach the limit of fail count as a client.
+	if clientErrCount >= config.FailCountLimit {
+		logrus.Errorf("srcCID:%s taskID:%d reach error limit", clientID, taskID)
+		return
+	}
+
+	dstCID = ps.tryGetCID(blackInfo, serviceErrorInfo, serviceDownInfo, producerLoad)
+	return
+}
+
+// tryGetCID returns a available dstCID from ps.pieceContainer.
+// Returns empty if no peer is available.
+func (ps *pieceState) tryGetCID(blackInfo map[string]bool, serviceErrorInfo, serviceDownInfo, producerLoad cutil.SyncMap) string {
+	ps.Lock()
+	defer ps.Unlock()
+
+	for k := range ps.pieceContainer {
+
+		// if the service of k has been down, and then it should not be needed.
+		if _, err := serviceDownInfo.Get(k); err == nil {
+			delete(ps.pieceContainer, k)
+			continue
+		}
+
+		// if service of k has failed for EliminationLimit times, and then it should not be needed.
+		if v, err := serviceErrorInfo.GetAsInt(k); err == nil &&
+			v >= config.EliminationLimit {
+			delete(ps.pieceContainer, k)
+			continue
+		}
+
+		// if the k is in the blackList, try the next one.
+		if isExistInMap(blackInfo, k) {
+			continue
+		}
+
+		load, err := producerLoad.GetAsInt(k)
+		if err != nil {
+			// if failed to get producer load, and then it should not be needed.
+			delete(ps.pieceContainer, k)
+			continue
+		}
+
+		if load < config.PeerUpLimit {
+			producerLoad.Add(k, load+1)
```

Comment:
Here's not threads-safety because that one task has one or more pieces and one peer has one or more tasks.